### PR TITLE
Resolving parent classes from other dirs/php internal classes [WIP]

### PR DESCRIPTION
**Pull request by vlastavesely**
Just a concept for a review. :construction: 

Tests are failing but it generates what I expected. :ok_hand: 

Related to #942

### CODE REVIEW

File: packages/Reflection/src/Parser/Parser.php [hunk 1]
```diff
@@ -127,11 +127,44 @@ private function transformBetterFunctionReflections(FunctionReflector $functionR
      */
     private function transformBetterClassInterfaceAndTraitReflections(ClassReflector $classReflector): array
     {
-        $betterClassReflections = $classReflector->getAllClasses();
+        $betterClassReflections = $this->resolveParentClassesAndInterfaces($classReflector->getAllClasses());
```
Comment (TomasVotruba) [inline packages/Reflection/src/Parser/Parser.php]: I'd split this into 2 lines.

File: packages/Reflection/src/Parser/Parser.php [hunk 2]
```diff
@@ -127,11 +127,44 @@ private function transformBetterFunctionReflections(FunctionReflector $functionR
...
 
         return $this->transformerCollector->transformGroup($betterClassReflections);
     }
 
+    /**
+     * @param ReflectionClass[]
+     */
+    private function resolveParentClassesAndInterfaces(array $betterClassReflections): array
```
Comment (TomasVotruba) [inline packages/Reflection/src/Parser/Parser.php]: Include "Traits" in the method name.

File: packages/Reflection/src/Parser/Parser.php [hunk 3]
```diff
@@ -127,11 +127,44 @@ private function transformBetterFunctionReflections(FunctionReflector $functionR
...
+    {
+        $allReflections = [];
+
+        foreach ($betterClassReflections as $class) {
+            $allReflections[$class->getName()] = $class;
+
+            foreach ($class->getInterfaces() as $interface) {
```
Comment (TomasVotruba) [inline packages/Reflection/src/Parser/Parser.php]: I'd split these into 3 methods.

File: packages/Reflection/src/Parser/Parser.php [hunk 4]
```diff
@@ -127,11 +127,44 @@ private function transformBetterFunctionReflections(FunctionReflector $functionR
...
 
+    /**
+     * @param ReflectionClass[]
+     */
+    private function resolveParentClassesAndInterfaces(array $betterClassReflections): array
+    {
+        $allReflections = [];
+
+        foreach ($betterClassReflections as $class) {
+            $allReflections[$class->getName()] = $class;
```
Comment (TomasVotruba) [inline packages/Reflection/src/Parser/Parser.php]: What about removing this line and using directorly `$betterClassReflections`? It includes already all found classes.

**Comment by TomasVotruba** [comment]
Just few details, otherwise it looks good :+1: 

Could you add tests for final version that fail before this fix? So we know this fixes it. E.g. similar use case #942

File: packages/Reflection/src/Parser/Parser.php (hunk 4 above)
Comment (vlastavesely) [inline packages/Reflection/src/Parser/Parser.php]: It includes but the items are identified by numeric index and not class name. So duplicities can occur.

Comment (TomasVotruba) [inline packages/Reflection/src/Parser/Parser.php]: I see. It makes sense then.

**Comment by vlastavesely** [comment]
The `TraitReflectionTest` is failing for BR cannot find `ParentTrait` in `SimpleTrait`. No test does any assertion on it so it could be removed. But it would be better to find source of problems and fix it properly. 

@TomasVotruba could you take a look on it? You may see something I missed.

**Comment by vlastavesely** [comment]
It is the Roave/BetterReflection#274 I think.

**Comment by TomasVotruba** [comment]
@vlastavesely I recall this issue now. You can either remove it or send failing PR, that @asgrim mentions in [BetterReflection comment](https://github.com/Roave/BetterReflection/pull/274#issuecomment-317268219) into `phpDocumentor/TypeResolver` package.

**Comment by vlastavesely** [comment]
I have a few things in process, so I should not start with another :) I commented it out so it can be easily added back.

**Comment by TomasVotruba** [comment]
:+1: thank you

**Pull request merged**